Add support for magnum

mbanszex opened the pull request:
Change-Id: I00041cc09ecd65ea82b15c532d9dba0669659663
Signed-off-by: Martin Banszel <martinx.banszel@intel.com>

mbanszex committed:
Add support for magnum

Signed-off-by: Martin Banszel <martinx.banszel@intel.com>

rbbratta commented:
It seems there are several flake8 and other errors.  Are you able to find them in the travis-ci output?  It might be too verbose.

mbanszex commented:
they are plenty. Seems that the settings is different from what we had in gerrit.

rbbratta commented:
flake8 and pylint are passing.   We output all the flake8, pylint warning and only a certain subset are considered fatal.

I think right now it is the unittests that are failing.

rbbratta commented:
All are all the unittest failures related to the pull request?  Because in this case, ensuring unittests pass, the standard procedure is to squash.  The general principal being that unittests should pass for every commit in the git history.  Or in other words, no commit in the git history should fail unittests.  This allows for git bisect and other tools to work.

If the unittest failures are in other parts of the code then, it would make sense to split those fixes into a different commits since they aren't directly related to this change.

esmacgil commented on taf/testlib/linux/service_lib.py:
```diff
@@ -184,20 +234,48 @@ def __init__(self, service_command_generator, run):
+_return_codes = {"systemd": SystemdReturnCodes}
+
 
 def specific_service_manager_factory(service_name, run_func):
```
These "factory" methods should be merged into their respective `__init__` methods.

mbanszex replied:
Done.

esmacgil commented on taf/testlib/magnum.py:
```diff
@@ -0,0 +1,145 @@
+DISTRO_METADATA = {
+    'fedora-atomic': {
+        'ssh_user': 'fedora',
+        'ssh_port': 22
```
Please include the trailing comma for literals when they are formatted in this manner.

mbanszex replied:
Done.

esmacgil commented on taf/testlib/magnum.py:
```diff
@@ -0,0 +1,145 @@
+
+        template.setdefault('external_network_id', self.config.network.public_network_id)
+        template.setdefault('keypair_id', self.venv.key['name'])
+        for a_default in CLUSTER_TEMPLATE_DEFAULTS:
```
When keys and values of the same dictionary are needed within a loop, please iterate over `.items()` instead of only `.keys()`.

But here a `ChainMap` would be yet better.

esmacgil commented on taf/testlib/rrdtool_graph.py:
```diff
@@ -97,31 +94,31 @@
+    ],
     'INTERFACE': [
         FileInfo('if_octets.rrd', 'incoming', 'rx'),
         FileInfo('if_octets.rrd', 'outgoing', 'tx')
```
Please add trailing commas too.

NazarTkachuk commented on taf/testlib/linux/bench.py:
```diff
@@ -0,0 +1,247 @@
+"""
+@copyright Copyright (c) 2015-2016, Intel Corporation.
+
```
Incorrect license. Need to update on Apache v2.0

NazarTkachuk commented on utils/ab_parser.py:
```diff
@@ -0,0 +1,214 @@
+"""
+@copyright Copyright (c) 2015-2016, Intel Corporation.
+
+This program is free software; you can redistribute it and/or modify it
```
Incorrect license. Need to update on Apache v2.0

NazarTkachuk commented on taf/testlib/linux/etcd_helper.py:
```diff
@@ -0,0 +1,100 @@
+import itertools
```
Add copyright statement

NazarTkachuk commented on taf/testlib/linux/kubernetes.py:
```diff
@@ -0,0 +1,37 @@
+from k8sclient.client import api_client
```
Add copyright statement

NazarTkachuk commented on taf/testlib/linux/utils.py:
```diff
@@ -0,0 +1,45 @@
+import time
```
Add copyright statement

NazarTkachuk commented on utils/images/onp_benchmark/Dockerfile:
```diff
@@ -0,0 +1,21 @@
+FROM fedora/nginx
```
Add line 
MAINTAINER intel.com

NazarTkachuk commented on utils/images/onp_benchmark/invoke_test.py:
```diff
@@ -0,0 +1,547 @@
+"""
```
Incorrect license

NazarTkachuk commented on utils/images/onp_benchmark/Dockerfile:
```diff
@@ -0,0 +1,21 @@
+FROM fedora/nginx
+ENV http_proxy http://proxy.jf.intel.com:911
+RUN echo "proxy=http://proxy.jf.intel.com:911" >> /etc/dnf/dnf.conf
```
we don't share internal domain names and ip adresses

NazarTkachuk commented on utils/images/onp_benchmark/nginx.conf:
```diff
@@ -0,0 +1,89 @@
+# For more information on configuration, see:
```
Add copyright statement

rbbratta replied:
no copyright on config data

mbanszex committed:
Merge service_lib factory methods into their respecitve __init__ methods

Signed-off-by: Martin Banszel <martinx.banszel@intel.com>

mbanszex committed:
Fixing licenses and formatting

Signed-off-by: Martin Banszel <martinx.banszel@intel.com>

rbbratta commented:
Fixed copyright to 2016 - 2017.  Remember when editing files to update copyright to 2017.

mbanszex committed:
Move some methods from a helper module in testcases to taf

Signed-off-by: Martin Banszel <martinx.banszel@intel.com>

mbanszex committed:
Fix unittests SpecificServiceManager constructor call

Signed-off-by: Martin Banszel <martinx.banszel@intel.com>

mbanszex committed:
Simplify service_lib and its moduletests

Support only systemd now

Signed-off-by: Martin Banszel <martinx.banszel@intel.com>

mbanszex committed:
Add get_services to virtual_env

Signed-off-by: Martin Banszel <martinx.banszel@intel.com>

esmacgil commented on taf/plugins/pytest_sut_monitor.py:
```diff
@@ -348,6 +354,9 @@ class SutMonitorPlugin(object):
     """
 
+    def __init__(self):
+        self.sut_monitor = None
```
Call super please

mbanszex replied:
Done

esmacgil replied:
Why don't I see a call to super?

esmacgil commented on taf/plugins/pytest_sut_monitor.py:
```diff
@@ -359,7 +368,7 @@ def monitor_init(self, env_init):
 
     @pytest.fixture(scope=setup_scope(), autouse=True)
-    def monitor(self, request, env_main, monitor_init):
+    def monitor(self, request, env_main, monitor_init):  # pylint: disable=W0613
```
Is it possible to change our pylint so that this isn't a stopper?  I like that pylint tells us about this, but I would prefer that it not be a blocking error.

rbbratta replied:
add it here https://github.com/taf3/taf/blob/ci/project_checker.py#L578

rbbratta replied:
err it's not in the fatal list, so pylint won't block on W0613.  So it is acting just as a warning currently.

https://github.com/taf3/taf/blob/ci/project_checker.py#L958

esmacgil commented on taf/testlib/linux/bench.py:
```diff
@@ -0,0 +1,290 @@
+    CLASS_LOGGER = loggers.ClassLogger()
+
+    def __getattr__(self, key):
+        try:
```
This is what suppress was made for and please include the attribute name in the Attribute exception:

```
with suppress(KeyError):
  return self._config[key]
raise AttributeError(key)
```

mbanszex replied:
Done.

esmacgil commented on taf/testlib/linux/bench.py:
```diff
@@ -0,0 +1,290 @@
+            raise AttributeError
+        self._config = ChainMap(config, self._ATTRIBUTES)
+
+    def set_labels(self, mode):
```
This looks like a mapping lookup function:

```
try:
  server_file, client_file, label = MODE_MAPPING[mode]
except KeyError:
  raise Exception(...)

set_node_selector(server_file, label)
set_node_selector(client_file, label)
```

esmacgil commented on taf/testlib/linux/bench.py:
```diff
@@ -0,0 +1,290 @@
+        self._config = ChainMap(config, self._ATTRIBUTES)
+
+    def set_labels(self, mode):
+        if mode == 'single_vm':
```
Please define a constant for these String literals.

esmacgil commented on taf/testlib/linux/bench.py:
```diff
@@ -0,0 +1,290 @@
+
+    @property
+    def test_type(self):
+        if self._test_type is None:
```
Since we aren't unifying the return statements, I'd prefer this to be the short path:

```
if self._test_type is not None:
  return self._test_type
client_file = ...
...
```

thus allowing the de-indentation of the body of this function.

rbbratta commented:
Needs review

esmacgil commented on taf/testlib/linux/etcd_helper.py:
```diff
@@ -0,0 +1,118 @@
+
+    @property
+    def latest_id(self):
+        try:
```
Please use suppress:

```
with suppress(AttributeError):
  return self._latest_id
with suppress(Exception):
  for _ in range(2):
    with suppress(etcd.EtcdKeyNotFound):
      self._latest_id = ...
      return self._latest_id
    self.init_etcd()
raise EtcdHelperException(...)
```

rbbratta replied:
why are we suppressing Exception?  I thought we implemented something that had an exclude list so we don't suppress SyntaxError, NameError, etc.  

No code should ever catch just Exception.

mbanszex replied:
I used the Mac's suppress version leaving out the supress(Exception). If something unknown happens it should drop down to the caller and the exception call stack should not be overwritten with calling raise.

esmacgil commented on taf/testlib/rrdtool_graph.py:
```diff
@@ -304,8 +301,8 @@
                (FILES['DISK'][1].vname, CDEF[1], VARS[3])], '%8.1lf%S {}')
-        ]
-    }
+    ]
```
This should have a comma, did flake8-comma not flag this?

esmacgil commented on taf/testlib/rrdtool_graph.py:
```diff
@@ -170,59 +167,59 @@
     Variable('MINIMUM', 'Min',),
     Variable('AVERAGE', 'Avg,'),
     Variable('MAXIMUM', 'Max,'),
     Variable('LAST', 'Last\l')
```
Another missing comma

esmacgil commented on taf/testlib/rrdtool_graph.py:
```diff
@@ -170,59 +167,59 @@
         LineDef(FILES['DISK'][1].vname, CDEF[2], True, False, 'Written\:'),
-        ]
-    }
+    ]
```
Missing comma

mbanszex committed:
Turn constants into iterable classes of constants.

- IterableMetaclass and it's class Constants introduced
  (in addition to group constants it allows to iterate over
   constants' values)
- reiview comments addressed

rbbratta commented:
Review scrub: needs review

mbanszex committed:
Improve test type detection

rbbratta commented:
Mac to review again

esmacgil commented on taf/testlib/linux/bench.py:
```diff
@@ -0,0 +1,308 @@
+                client_file[test_type]  # pylint: disable=W0104
+                self._test_type = test_type
+                return self._test_type
+            raise BenchmarkException('{}: Unknown test type'.format(test_type))
```
Is this correct?  Am I reading this correctly: that if the first `test_type` in `self.PARSERS` is not in `client_file` then the exception will be raised.  Shouldn't the `raise` statement be at the same indentation as the `for` statement?

Also, since this code doesn't need the value at `client_file[test_type]`, we can test for the presence instead of doing the look up:

```
for test_type in self.PARSERS:
    if test_type in client_file:
        self._test_type = test_type
        return self._test_type
raise BenchmarkException(...)
```

or

```
with suppress(StopIteration):
    test_type = next(test_type for test_type in self.PARSERS if test_type in client_file)
    self._test_type = test_type
    return self._test_type
raise BenchmarkException(...)
```

mbanszex replied:
yes, that's a bug.

esmacgil commented on taf/testlib/tempest_clients/magnum/clients/cluster_client.py:
```diff
@@ -0,0 +1,175 @@
+                lambda: self.does_cluster_exist(cluster_id), 10, 1800)
+        except Exception:
+            # In error state.  Clean up the cluster id if desired
+            self.LOG.error('Cluster %s entered an exception state.' %
```
Do not use modulo (`%`) when calling logger methods.

`self.LOG.error('Cluster %s entered an exception state', cluster_id)`

mbanszex replied:
This magnum client code is taken over from magnum project. I wanted to keep it in the original shape. Except for some imports adjustments, I haven't touched it.

mbanszex committed:
Time out when cleaning Kubernetes resources takes too long

esmacgil commented on taf/testlib/virtual_env.py:
```diff
@@ -162,20 +180,63 @@ def __init__(self, opts=None):
+                with open(pubkeyfile, 'r') as stream:
+                    key['public_key'] = stream.read()
+            except IOError as e:
+                self.class_logger.info("ssh key provided but can't be used! ({})".format(e))
```
Logging with format.

mbanszex replied:
Fixed

esmacgil commented on taf/testlib/virtual_env.py:
```diff
@@ -310,17 +376,19 @@ def ensure_public_access(self, try_reuse=False, networks_client=None, routers_cl
+                'enable_snat': True
+            }
+            public_router = self.create_router(**router_kwargs)
+            assert public_router
```
Won't the next statement raise some exception if `public_router` is not a dictionary or has no values?

The only thing this statement seems to protect against , beyond the statement that follows it, is an _empty_ `defaultdict`, while it would pass through many other invalid values: `True`, `1`, `[0]`, `object()` and many more.

mbanszex replied:
Yes, that's right. This assert is meant only to catch the situation when the cluster creation fails and the tempest returns None. We could test a lot of things to make sure that we got a correct dictionary. But if it woudn't have e.g. the id key, something would be very wrong and then I think the KeyError exception or similar is appropriate.

esmacgil commented on taf/testlib/virtual_env.py:
```diff
@@ -310,17 +376,19 @@ def ensure_public_access(self, try_reuse=False, networks_client=None, routers_cl
+            _net_cfg.public_router_id = public_router['id']
 
-        if _net_cfg.public_network_id and _net_cfg.public_router_id:
+        if _net_cfg.public_network_id and (_net_cfg.public_router_id or
```
Please return the result of the Boolean expression instead of testing the Boolean expression and returning a Boolean literal.

```
return _net_cfg.public_network_id and (...)
```

mbanszex replied:
Done

mbanszex committed:
Add trailing commas and improve SFC exception handling in virtual_env.py

mbanszex committed:
Add clear_labels method to KuberetesHelper

Signed-off-by: Martin Banszel <martinx.banszel@intel.com>

esmacgil commented:
I see you say that comments have been addressed, but where is the latest change set?

mbanszex committed:
Review comments

Signed-off-by: Martin Banszel <martinx.banszel@intel.com>

mbanszex commented:
Sorry, I somehow lost part of the changes. Added now. Before merging to master, we need to squash all the commits either into one or refactor the commits so that they make some sense. Currently they just reflect the review process, not the logical changes.

esmacgil commented on taf/testlib/virtual_env.py:
```diff
@@ -304,26 +371,24 @@ def ensure_public_access(self, try_reuse=False, networks_client=None, routers_cl
+            _net_cfg.public_router_id = public_router['id']
 
-        return False
+        return  _net_cfg.public_network_id and (_net_cfg.public_router_id or not create_external_router)
```
Looking at this again, would it make more sense to put the `not create_external_router` clause before `_net_cfg.public_router_id`?  Because the latter could raise an AttributeError even when the result should be True.

```return _net_cfg.public_network_id and (not create_external_router or _net_cfg.public_router_id)```

mbanszex replied:
Done

mbanszex committed:
Add a helper action-cleanup function for fixtures

mbanszex committed:
Change in condition evaluation

mbanszex committed:
Add unittests for ab_parser

Signed-off-by: Martin Banszel <martinx.banszel@intel.com>

esmacgil commented on taf/plugins/pytest_sut_monitor.py:
```diff
@@ -227,20 +229,23 @@ def list_rrd_folders(self):
-            folders = self.exec_command('find {}/*'
-                ' -maxdepth 1 -type d -print0'.format(collectd_folder)).split('\0')
+            folders = self.exec_command(
+                'find {}/*' ' -maxdepth 1 -type d -print0'.format(collectd_folder)).split('\0')
```
can we make this command a list instead of a string?

```
folders = self.exec_command(['find', '{}/*'.format(collectd_folder), '-maxdepth', '1', '-type', 'd', '-print0']).split('\0')
```

mbanszex replied:
LinuxHostBash.cli_send_command() and paramiko exec_command() accept a string, not a list of strings.
List might be better but I don't want to change this in the whole TAF in this PR. We can change that in a separate PR.

esmacgil commented on taf/testlib/linux/service_lib.py:
```diff
@@ -133,71 +134,133 @@ def method(service_name):
+            kwargs.pop('service', None)  # remove any value associated with the service key
+            return self.run_func(command, **kwargs)
+        command = getattr(self.service_command_generator, name)
+        command = ' '.join(command(self.service_name))
```
is there a reason to do this join when the command execution function should do it for us?

esmacgil commented on taf/testlib/linux/service_lib.py:
```diff
@@ -133,71 +134,133 @@ def method(service_name):
+
+    def __getattr__(self, name):
+        def run(service='', **kwargs):
+            return self.run_func(' '.join(command(service)), **kwargs)
```
is there a reason to do this join when the command execution function should do it for us?

mbanszex replied:
The run_func is given by the caller. At the point all exec_command or cli_send_command implentations in TAF expect the command to be a string. Sometimes thy even call command.split(' ') -- which is totaly wrong. (taf/testlib/clinns.py:174)
This should be fixed globaly in a separate PR.

mbanszex committed:
Add missing super().__init__() call to the pytest_sut_monitor.py
